cxx-qt-gen/src/parser/mod.rs

```diff
@@ -0,0 +1,166 @@
+// SPDX-FileCopyrightText: 2022 Klarälvdalens Datakonsult AB, a KDAB Group company <info@kdab.com>
+// SPDX-FileContributor: Andrew Hayzen <andrew.hayzen@kdab.com>
+//
+// SPDX-License-Identifier: MIT OR Apache-2.0
+
+pub mod qobject;
+
+use qobject::{find_and_merge_cxx_qt_item, find_qobjects, ParserCxxQtData};
+use syn::{token::Brace, ItemMod, Result};
+
+/// A struct representing a module block with CXX-Qt relevant Items into ParserCxxQtData
+pub struct Parser {
+    /// The module which unknown (eg CXX) blocks are stored into
+    pub module: ItemMod,
+    /// Any CXX-Qt data that needs generation later
+    pub cxx_qt_data: ParserCxxQtData,
+}
+
+impl Parser {
+    /// Constructs a Parser object from a given ItemMod block
+    pub fn from(module: ItemMod) -> Result<Self> {
+        let mut new_module = module.clone();
+        let mut cxx_qt_data = ParserCxxQtData::default();
+        let mut others = vec![];
+
+        // Check that there are items in the module
+        if let Some(mut items) = module.content {
+            // Find any QObject structs
+            cxx_qt_data.qobjects = find_qobjects(&items.1)?;
+
+            // Loop through any qobjects that were found
+            if !cxx_qt_data.qobjects.is_empty() {
+                for item in items.1.drain(..) {
+                    // Try to find any CXX-Qt items, if found add them to the relevant
+                    // qobject. Otherwise return them to be added to other
+                    if let Some(other) = find_and_merge_cxx_qt_item(item, &mut cxx_qt_data)? {
+                        // Unknown item so add to the other list
+                        others.push(other);
+                    }
+                }
+            } else {
+                // No qobjects found so pass everything through
+                others.extend(items.1);
+            }
+        }
+
+        // Create a new module using only items that are not CXX-Qt items
+        new_module.content = Some((Brace::default(), others));
+
+        // Return the successful Parser object
+        Ok(Self {
+            module: new_module,
+            cxx_qt_data,
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use quote::{quote, ToTokens};
+    use syn::ItemMod;
+
+    #[test]
+    fn test_parser_from() {
+        // Empty module
+        let tokens = quote! {
+            #[cxx_qt::bridge]
+            mod ffi {}
+        }
+        .into_token_stream();
+        let module: ItemMod = syn::parse2(tokens).unwrap();
+        let parser = Parser::from(module);
+        assert!(parser.is_ok());
+        let parser = parser.unwrap();
+        assert_eq!(parser.module.attrs.len(), 1);
+        assert_eq!(parser.module.ident, "ffi");
+        assert_eq!(parser.module.content.unwrap().1.len(), 0);
```

Comment:
Could these tests assert that the passed through module is identical to a specified `syn::ItemMod` rather than testing a few pieces of metadata about the module?